openeo_driver/datacube.py

```diff
@@ -770,9 +771,44 @@ def apply_dimension(
                 or (dimension in {self.DIM_BANDS, self.DIM_PROPERTIES}.intersection(self.get_dimension_names()))
                 and target_dimension is None
             ):
+
+                def check_udf_result(result: openeo.udf.UdfData, require_datacube=False):
+                    if not isinstance(result, openeo.udf.UdfData):
+                        raise ValueError(f"UDF should return UdfData, but got {type(result)}")
+                    feature_list = result.get_feature_collection_list()
+                    if not (feature_list and len(feature_list) == 1):
+                        raise ValueError(
+                            f"UDF should return single feature collection but got {feature_list and len(feature_list)}"
+                        )
+                    if require_datacube:
+                        datacube_list = result.get_datacube_list()
+                        if not (datacube_list and len(datacube_list) == 1):
+                            raise ValueError(
+                                f"UDF should return single datacube but got {datacube_list and len(datacube_list)}"
+                            )
+                        return feature_list[0].data, datacube_list[0].array
+                    return feature_list[0].data, None
+
                 log.warning(
                     f"Using experimental feature: DriverVectorCube.apply_dimension along dim {dimension} and empty cube"
                 )
+                if len(re.findall(r"^def apply_vectorcube\(", single_run_udf.udf, re.MULTILINE)) != 0:
```

Comment:
so this should be more to the point (with some additional whitespace paranoia):
```suggestion
                if re.search(f"^def\s+apply_vectorcube\s*\(", single_run_udf.udf, re.MULTILINE):
```